### PR TITLE
feat: benchmark

### LINKED ISSUE #917: Built-in benchmarking support
### Clear and concise description of the problem

Since we could write unit test in source files now with vitest I wish to also write some benchmarks like we do in rust/go. ~Deno cli will support bench command soon (https://github.com/denoland/deno/pull/13713)~ Deno added `deno bench` subcommand recently in [1.20.0 release](https://deno.com/blog/v1.20#new-subcommand-deno-bench).

In Node.js most of the project uses a library called [benchmark](https://www.npmjs.com/package/benchmark) whose last commit was 5 years ago. It works pretty well honestly but things changed a lot since then especially we are now suffering from the pain of esm/cjs interop. With vitest it could provide out-of-box ESM or even TypeScript support to bring a much more friendly dx.

### Suggested solution

Rust uses bench attribute and `--bench` flag should be passed in or just `cargo bench`  to run it:

```rust
// function to benchmark must be annotated with `#[bench]`
#[bench]
fn recursive_fibonacci(b: &mut Bencher) {
    // exact code to benchmark must be passed as a closure to the iter
    // method of Bencher
    b.iter(|| {
        (0..BENCH_SIZE).map(fibonacci).collect::<Vec<u32>>()
    })
}

#[bench]
fn iterative_fibonacci(b: &mut Bencher) {
    b.iter(|| {
        fibonacci_sequence().take(BENCH_SIZE).collect::<Vec<u32>>()
    })
}
```

Benchmarking is similar to unit testing in Golang, run it with `go test -bench=.`:

```go
// from fib_test.go
func BenchmarkFib10(b *testing.B) {
        // run the Fib function b.N times
        for n := 0; n < b.N; n++ {
                Fib(10)
        }
}
```

In deno, see https://github.com/denoland/deno/pull/13713:

```js
// Defaults to 1000 warmup iterations and 1000 bench iterations
Deno.bench("URL parsing", { n: 250, warmup: 100 }, () => {
  new URL("https://deno.land/");  
});


Deno.bench("URL resolving", { n: 1000 }, () => {
  new URL("./foo.js", import.meta.url);
});
```

```sh
$ deno bench --unstable example_bench.ts
running 2 benches from file:///Users/ib/dev/deno/example_bench.ts
bench URL parsing ... 250 iterations 23,063 ns/iter (208..356,041 ns/iter) ok (1s)
bench URL resolving ... 1000 iterations 48,045 ns/iter (250..128,625 ns/iter) ok (4s)

bench result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (5s)
```

My first thought was to provide a `benchmark` function, it works just like `describe` except it only runs with a bench flag:

```js
import { benchmark } from 'vitest'

benchmark('fib', (bench) => {
  bench('recursive', () => {
    recursiveFibonacci()
  })
  bench('iterative', () => {
    iterativeFibonacci()
  })
})
```

Also the output could be much prettier and modern comparing to `benchmark` package.

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by poyoho**
fix #917

native support benchmark.

* Create a new `bench` command to start vitest and distinguish the current running mode of vitest (unit test or benchmark).
* Reuse the Vitest suite collection benchmark, add the `bench` API to distinguish between unit test and benchmark and collect the benchmark execution function.
* When running the test, unit-test normally enters the unit-test branch and executes the unit-test function once. And benchmark will use `Tinybench@2.1.3` as engine asynchronous multiple execution function.
* provide verbose and json reporter to show the benchmark result.

### CODE REVIEW

File: packages/vitest/package.json [hunk 1]
```diff
@@ -85,6 +85,7 @@
   "dependencies": {
     "@types/chai": "^4.3.0",
     "@types/chai-subset": "^1.3.3",
+    "benchmark": "^2.1.4",
```
Comment (antfu) [inline packages/vitest/package.json:88]: I haven't looked close to it but seems this package is no longer maintained for 5 years already. Do we have a better alternative?

Comment (poyoho) [reply packages/vitest/package.json:88]: I saw the issues recommendation and didn't look for other 🙈

Comment (poyoho) [reply packages/vitest/package.json:88]: @antfu There seems to be no better alternative https://www.npmjs.com/search?q=keywords:benchmark. I think it is possible to fork `benchmark` to maintain a new one?

**Comment by antfu** [comment]
Let's discuss the design in #917 first.

About benchmark.js, if we really end up using it, I could try to contact the author and see if our team could grant the access and help on maintaining it.

File: packages/vitest/package.json (hunk 1 above)
Comment (thelinuxlich) [reply packages/vitest/package.json:88]: There is a fork by @Uzlopak actively updated in https://github.com/cthulhu-node/benchmark.js might be worthy joining forces

Comment (Uzlopak) [reply packages/vitest/package.json:88]: I am now back from a three weeks vacation. I still want to implement the ability to automatically detect async functions and add the one PR to the original project which was fixing the statistics.
I did not publish a package yet. 
But we can surely join forces. :)

**Comment by poyoho** [comment]
It seems need to make benchmark(`xxxx.bench.js`) whit diff files and add a new command to run benchmark 👏

**Comment by sheremet-va** [comment]
I think we can now use [tinybench](https://github.com/tinylibs/tinybench) by @Uzlopak?

**Comment by Uzlopak** [comment]
Published version 1.0.2 of tinybench.

**Comment by poyoho** [comment]
image 
😁

**Comment by Uzlopak** [comment]
Created a PR against github-action-benchmark. Does the color have some effect on the stdout and parsing by benchmark.js?

https://github.com/benchmark-action/github-action-benchmark/pull/126

File: test/benchmark/test/benchmark.test.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+describe('sort', () => {
+  benchmark('normal', () => {
+    const x = [1, 5, 4, 2, 3]
+    x.sort((a, b) => {
+      return a - b
+    })
+  })
+
+  benchmark('reverse', () => {
```
[the remark this answers is not in the document]
Comment (poyoho) [reply test/benchmark/test/benchmark.test.ts:11]: oh, you make the `fn` to be promise?

[the remark this answers is not in the document]
Comment (Uzlopak) [reply test/benchmark/test/benchmark.test.ts:11]: There is a wrapper for async functions in tinybench.

This code snippet just an example to see that tinybench is not going crazy and calling the async function a gazillion times. So you should see that async makes about 100 ops/s. 

Theoretically also your reverse example could `benchmark('reverse', async () => {` and should get also about 5.6 mio. ops/s.

File: packages/vitest/src/types/benchmark.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  hz: number
+  rme: number
+  sampleSize: number
+}
+
+export type BenchmarkOptions = TinyBench.Options
+export type BenchFunction = (this: TinyBench.Suite, deferred: TinyBench.Deferred) => void
+export type BenchmarkAPI = ChainableFunction<
+'skip',
+[name: string, fn?: BenchFunction, options?: BenchmarkOptions],
```
Comment (Uzlopak) [inline packages/vitest/src/types/benchmark.ts:55]:
```suggestion
[name: string, fn: BenchFunction, options?: BenchmarkOptions],
```

Why is fn optional? We need always a benchmark function

**Comment by Uzlopak** [comment]
I realized that the async functionality is not working well.

File: packages/vitest/src/runtime/run.ts [hunk 4]
```diff
@@ -237,10 +242,89 @@ export async function runSuite(suite: Suite) {
...
+    benchmarkLib.on('complete', () => {
+      suite.result!.duration = performance.now() - start
+      suite.result!.state = 'pass'
+      updateTask(suite)
+      defer.resolve(null)
+    })
+    benchmarkLib.on('error', (e) => {
+      defer.reject(e)
+    })
+    benchmarkLib.run({ async: true })
```
Comment (Uzlopak) [inline packages/vitest/src/runtime/run.ts:318]: also add `delay: 0`

Comment (poyoho) [reply packages/vitest/src/runtime/run.ts:318]: `benchmarkLib.run({ async: true, delay: 0 })` ?

Comment (Uzlopak) [reply packages/vitest/src/runtime/run.ts]: yeah

**Comment by mangs** [comment]
Hmm, looking at the screenshot, average time per run is not mentioned, there is no breakdown by percentile, and which test was fastest is not mentioned; this can be important if you're comparing 4+ things. Can that be added?

I strongly suggest taking a look at [mitata](https://github.com/evanwashere/mitata) before this PR gets merged. This is how it looks integrated into [Deno](https://deno.land/) (it's what Deno uses for the `deno bench` command): 
image

File: packages/vitest/src/node/cli-api.ts [hunk 5]
```diff
@@ -24,12 +24,17 @@ export async function startVitest(cliFilters: string[], options: CliOptions, vit
...
+  if (typeof options.coverage === 'boolean' && !options.benchmark)
     options.coverage = { enabled: options.coverage }
 
   const ctx = await createVitest(options, viteOverrides)
 
-  if (ctx.config.coverage.enabled) {
+  process.env.VITEST_MODE = ctx.config.watch ? 'WATCH' : 'RUN'
+
+  if (ctx.config.env)
+    Object.assign(process.env, ctx.config.env)
```
Comment (antfu) [inline packages/vitest/src/node/cli-api.ts:35]: These lines have been moved to pool.ts

**Comment by antfu** [comment]
I guess we should log an experiment warning when ppl use this command, addressing it "API might change in the future"

**Comment by poyoho** [comment]
@mangs you mean deno use mitata?

**Comment by mangs** [comment]
Hi @poyoho. The percentiles show outliers which can be important when tracking down performance issues which can get masked by averages. This is a very contrived example, but if you have 11 runs showing an average of 5 seconds, you can get that result with five items being zero, five 10, and one 5; percentiles would make that very clear. I think the function you found would do the trick and would be great! Thank you for considering my idea. As a big fan of what this project is doing, I don't mean to be a bother, I just want this feature to be as good as it can be. 😀

**Comment by sheremet-va** [comment]
Sorry, I noticed with `rebase` came a lot of `@jest` packages. Is this intentional?

**Comment by poyoho** [comment]
image

I think now log data is more, and the detailed data need to log in another way. so the pxx data I want to report in the file. 😁

**Comment by Uzlopak** [comment]
Thank you for the link to the mitata repo. I analyzed the code and tbh I think their code is much more cleaner than tinybench/benchmark.js can ever be. Maybe we should really consider switching to mitata. Also I assume that as it is integrated into deno it will have much more traction in the development than tinybench will ever be.

**Comment by Uzlopak** [comment]
@evanwashere

Maybe you want to add your thoughts to this feature?

**Comment by poyoho** [comment]
@Uzlopak so Can I switch the tinybench to mitata ?

**Comment by sheremet-va** [comment]
Do we have news on this PR? @Aslemammad you said, you wanted to take a look?

**Comment by Aslemammad** [comment]
I'll take a look, I'm trying to implement a typesafe tinybench and modern API for it (based on the current)! and see if I can something out of it!

@poyoho could you please let me know what you like about other testing libs over tinybench, I'd like to add those features in the new huge refactor.

**Comment by poyoho** [comment]
@Aslemammad sure! I think tinybench can export some runtime hook, so we can make the log to be  beautiful and show more runtime info. And mitata is seems just run the `fn` and show the benchmark info. 

tinybench can calculate many benchmark results based on his exposed hook. some like `mitata` I think we can even make it in vitest. It look like very easy 😁

**Comment by Aslemammad** [comment]
@poyoho That's great, let's have a chat about it then. here's my discord: Aslemammad#6670

**Comment by Aslemammad** [comment]
@poyoho The latest version of tinybench is `2.x.x` and the `1.x.x` is for the previous version which was based on benchmark.js.

**Comment by poyoho** [comment]
I know. Rebase first 😁

**Comment by poyoho** [comment]
image

@Aslemammad 🎉🎉🎉

**Comment by Aslemammad** [comment]
@poyoho Happy to see this result! Thank you so much for the work. how about changing the `benchmark` function name to `bench` (in the benchmark.test.ts)? That'd be an easier naming in my opinion.

**Comment by poyoho** [comment]
`benchmark` -> `bench`

File: packages/vitest/src/runtime/map.ts [hunk 6]
```diff
@@ -26,9 +25,13 @@ export function isTest(task: Test | Benchmark): task is Test {
   return task.type === 'test'
 }
 
-export function getBenchmarkFactory(key: Suite): BenchFactory {
+export async function getBenchmarkFactory(key: Suite): Promise<BenchFactory> {
   let benchmark = benchmarkMap.get(key)
   if (!benchmark) {
+    if (!globalThis.EventTarget)
+      await import('event-target-polyfill' as any)
```
Comment (Aslemammad) [inline packages/vitest/src/runtime/map.ts:32]: Please consider putting the `any` after the import call!

Comment (poyoho) [reply packages/vitest/src/runtime/map.ts:32]: putting the any after the import call can to fix the types error 🤦

**Comment by Uzlopak** [comment]
Just thought for improvements:

Consider running each benchmark in a separate process. So maybe it makes sense to recommend one benchmark per .bench.js/ts file. 

Also maybe consider adding a "benchmark normalizer". E.g. run first a benchmark calculating the fibonacci numbers and use the ops/s as 100%. So if you run the benchmarks on different machines, the benchmarks get more comparable.

**Comment by Aslemammad** [comment]
What I also suggest is that we should promote warming up, it can help with having better and more accurate benchmarks, which is easily possible with tinybench.

**Comment by Uzlopak** [comment]
Afaik warmup is done by default by benchmark.js. Maybe we should do it also automatically

**Comment by Aslemammad** [comment]
@Uzlopak I guess having as a function in Tinybench (as it is right now) is better, so folks can choose to do it or not!

**Comment by antfu** [comment]
I guess maybe we could have a `.bench.ts` convert to differentiated from testing

**Comment by poyoho** [comment]
done!

**Comment by antfu** [comment]
Pushed some changes,

- Introduce run mode. We should not rely on `config.benchmark` as it will present if users set the config regardless of running the command or not.
- Throw a hard error on calling `test()` in benchmark mode and vice versa. We could test that in `test/fails` later.

**Comment by antfu** [comment]
image

@poyoho the ops/sec are identical, there might be something wrong

**Comment by sheremet-va** [comment]
> image
> 
> @poyoho the ops/sec are identical, there might be something wrong

Even looking at this, it seems we show a lot less useful information than [Deno does](https://github.com/vitest-dev/vitest/pull/1029#issuecomment-1146744979)? (no percentile, for example) Didn't we want to take a look it this?

**Comment by Uzlopak** [comment]
I guess we need to calculate ttables and utables for p75 and p995

**Comment by Aslemammad** [comment]
@Uzlopak Tinybench already returns those percentiles https://github.com/tinylibs/tinybench#taskresult

**Comment by antfu** [comment]
We probably need to think more about how we present the result for sure.

**Comment by antfu** [comment]
We already support the reporter interface to access the full test result.

**Comment by poyoho** [comment]
In fact, we had the results, but this style can't fill too much data 😁

**Comment by poyoho** [comment]
> image
> 
> @poyoho the ops/sec are identical, there might be something wrong

two decimal places will be rounded off. The real result is:  0.006637481844709122, 0.005769233822295151...

**Comment by poyoho** [comment]
image
updated

**Comment by poyoho** [comment]
I think we can native support deno style benchmark result.

**Comment by antfu** [comment]
For ops/sec <1, should we display the duration of one op instead?

**Comment by poyoho** [comment]
maybe we can change ops/sec -> ops/ms

**Comment by Uzlopak** [comment]
I dont think, that ops/ms is that relevant. I personally run benchmarks multiple times because it also depends on which core you run the benchmark and if there is currently a heavier process running on it. So basically benchmarks only indicate if you have some heavy performance degredation. So I would think it is totally OK to have the ops/s. Also you have the percentage on how they differ, so you can actually see which one is faster.

Is it possible to pass options to the benchmark like

```js
describe('sort', { minSample: 100 }, () => {
  bench('normal', () => {
    const x = [1, 5, 4, 2, 3]
    x.sort((a, b) => {
      return a - b
    })
  })
})
```

**Comment by poyoho** [comment]
image

feat json format report to show all the result

**Comment by sheremet-va** [comment]
What do we need to release this?

**Comment by antfu** [comment]
I would like to improve the default output to a table, but not blocking as we could improve it later. Since this is experimental, I think we could ship it in the next minor.

File: packages/vitest/src/node/core.ts [hunk 7]
```diff
@@ -101,7 +103,9 @@ export class Vitest {
       })
     }
 
-    this.reporters = await createReporters(resolved.reporters, this.runner)
+    this.reporters = resolved.benchmark
```
Comment (sheremet-va) [inline packages/vitest/src/node/core.ts:106]: shouldn't this be decided based on `mode`?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/vitest/src/node/core.ts (hunk 7 above)
Comment (antfu) [reply packages/vitest/src/node/core.ts:106]: Nice catch!

Comment (sheremet-va) [reply packages/vitest/src/node/core.ts]: I don't think `resolved.mode` is the same as VitestMode (`'test' | 'benchmark'`), it's custom user mode (`--mode`)

Comment (sheremet-va) [reply packages/vitest/src/node/core.ts]: We now have three modes 😄 

- `'WATCH' | 'RUN'`
- `'test' | 'benchmark'`
- custom, default `test` - btw, should we change it to `benchmark`, if Vitest is running as `vitest bench`?

File: packages/vitest/src/runtime/run.ts [hunk 8]
```diff
@@ -230,23 +231,27 @@ export async function runSuite(suite: Suite) {
...
-        }
-        else {
-          const { sequence } = workerState.config
-          if (sequence.shuffle || suite.shuffle) {
-            // run describe block independently from tests
-            const suites = tasksGroup.filter(group => group.type === 'suite')
-            const tests = tasksGroup.filter(group => group.type === 'test')
-            const groups = shuffle([suites, tests], sequence.seed)
-            tasksGroup = groups.flatMap(group => shuffle(group, sequence.seed))
+      if (getRunMode()) {
```
Comment (sheremet-va) [inline packages/vitest/src/runtime/run.ts:234]: This looks weird 🤔 Shouldn't it be `isRunningInBenchmark`?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/vitest/src/node/core.ts (hunk 7 above)
Comment (sheremet-va) [reply packages/vitest/src/node/core.ts]: I see that now mode is always `'test' | 'benchmark'`, but it's still confusing, we will need to come up with better naming, I think :S

Still, I think we should pass down `benchmark` to `defineConfig` 🤔

File: packages/vitest/src/runtime/run.ts (hunk 8 above)
Comment (sheremet-va) [reply packages/vitest/src/runtime/run.ts]: Yep, all tests were skipped, and always passed :)

**Pull request merged**